### PR TITLE
PR-5975: Add support for optional keys

### PR DESCRIPTION
**Pull request by mckadesorensen**

### CODE REVIEW

File: mandible/metadata_mapper/format/format.py [hunk 1]
```diff
@@ -73,6 +73,8 @@ def _eval_key_wrapper(self, data, key: Key) -> Any:
         try:
             return self._eval_key(data, key)
         except KeyError as e:
+            if key.optional_key:
```
Comment (mckadesorensen) [inline mandible/metadata_mapper/format/format.py:76]: I was originally thinking about having the logic out of the except block, but after some thinking, I don't think that is the route I want to take. If I were to go that route each class that inherits from SimpleFormat would have to have that logic defined in self._eval_key.

**Review by gjclark — APPROVED** [review]

File: mandible/metadata_mapper/format/format.py (hunk 1 above)
Comment (mattp0) [reply mandible/metadata_mapper/format/format.py:76]: I do feel like this is better handled by each implementation of the SimpleFormat and using the try except feels wrong to me.

For example each of the systems have a `_eval_key()` defined for example in the jsonpath which json and zip use it seems a get_key() using the jsonpath is set up like a filter. if key.optional_key and not matches return a value.

https://github.com/asfadmin/mandible/blob/86e8c55f6c306057c659e8e7c357f465c4bdd190/mandible/jsonpath.py#L23

for the h5py I believe that groups could be leveraged here to find the keys and check if the optional key is there. https://docs.h5py.org/en/stable/high/group.html rather than the `data["key.key"]` that would lead to a `KeyError`
https://github.com/asfadmin/mandible/blob/86e8c55f6c306057c659e8e7c357f465c4bdd190/mandible/metadata_mapper/format/h5.py#L19

for the xml parse i believe is using the same filter idea from the xpath, returns a match or []. could use if key.optional_key and not matches return a value.
https://github.com/asfadmin/mandible/blob/86e8c55f6c306057c659e8e7c357f465c4bdd190/mandible/metadata_mapper/format/xml.py#L20

Comment (mckadesorensen) [reply mandible/metadata_mapper/format/format.py:76]: Generally, I agree. In this case, I don't think I do. If we were to introduce a new type, that is four locations that the logic is being defined. Not to mention, we have to remember to add it to that new type. If we want to change anything thing in the future, that is four cases we need to update.

Comment (mckadesorensen) [reply mandible/metadata_mapper/format/format.py:76]: From my understanding, even the Python core dev team will use exceptions for flow control. 
Screenshot 2024-08-08 at 4 34 10 PM

Comment (mckadesorensen) [reply mandible/metadata_mapper/format/format.py:76]: It isn't something that I think should be done often, in this case, I think it makes sense.

Comment (mattp0) [reply mandible/metadata_mapper/format/format.py:76]: The implementation of a data format needs to define the logic for what it means to handle an optional key.

Comment (mattp0) [reply mandible/metadata_mapper/format/format.py:76]: If the general consensus is this is a good way to do it then I will be fine.

Comment (mckadesorensen) [reply mandible/metadata_mapper/format/format.py:76]: Assuming there are cases that we need to handle it differently, we can always overwrite the method `_eval_key_wrapper` in that particular data format class. I think returning `None` when the 'optional_key` is set to True and there is no key, will be the solution most implementations need. Going this route, less code needs to be written and maintained.

Comment (reweeden) [reply mandible/metadata_mapper/format/format.py:76]: I think this is where I would put it. Feels like a good spot to me. Besides, the addition of this option check doesn't change anything about how the exceptions flow through the program. We need to catch and re-raise anyway because we need to be transforming the exception type.

**Review by mattp0 — APPROVED** [review]

File: mandible/metadata_mapper/exception.py [hunk 2]
```diff
@@ -1,3 +1,6 @@
+RAISE_EXCEPTION = object()
```
Comment (reweeden) [inline mandible/metadata_mapper/exception.py:1]: This should be defined in `key.py` as it exists exclusively to be the default value for the `default` argument of `Key`.

File: tests/test_format.py [hunk 3]
```diff
@@ -67,6 +67,17 @@ def test_h5():
...
 
+@pytest.mark.h5
+def test_h5_optional_key():
+    file = io.BytesIO()
+    with h5py.File(file, "w") as f:
+        f["foo"] = "foo value"
+
+    format = H5()
+
+    assert format.get_value(file, Key("bar", default=None)) is None
```
Comment (reweeden) [inline tests/test_format.py:78]: Should also add a test where the default is something other than `None`.

File: tests/test_format.py [hunk 4]
```diff
@@ -67,6 +67,19 @@ def test_h5():
     }
 
 
+@pytest.mark.h5
+def test_h5_optional_key():
```
Comment (reweeden) [inline tests/test_format.py:71]: Sorry, forgot to ask this earlier but could you also make sure that there are some tests for using the builder classes with a default value? I think everything should work without further modifications, but it would be good to add a test to confirm that.

I think it would be good to add a test in `test_builder.py` similar to `test_build_mapped_key_options` but with the default value.

And then it would also be good to update `integration_tests/test_full_example.py` as that is supposed to be basically a giant test covering every possible metadata mapper feature that we have.

Comment (mckadesorensen) [reply tests/test_format.py:71]: Added them.

**Review by reweeden — APPROVED** [review]
Excellent!
